Automattic/newspack-popups
# feat: lightweight GET API for campaigns

### All Submissions:






### Changes proposed in this Pull Request:

Refactors the endpoint that determines whether a Campaign should be shown as a lightweight, standalone application that uses WordPress' database and cacheing but doesn't incur the full weight of a WP REST API request. On a site with campaigns this endpoint will often be hit at least once for each anonymous page load, so the lighter and faster it is, the better. 

Set-up instructions are in https://github.com/Automattic/newspack-popups/blob/7e7f5df3a9a5057f6b6d5bb203fec63099591773/api/README.md. When testing on an Atomic site, the config can omit everything except for `DB_PREFIX`. 

To test cacheing locally, install `object-cache.php` from https://wordpress.org/plugins/memcached/ in `wp-content/`. 

### How to test the changes in this Pull Request:

1. Verify that the new endpoint is being used. If the config file is not found, the plugin will use the REST API endpoint instead.
2. Verify that all Campaign visibility and suppression flows work correctly.
3. With `memcached` installed in `wp-content/` and the `NEWSPACK_POPUPS_DEBUG` constant set to true in the config, observe the API responses. They will include information about the number of DB reads/writes and cache hits. When testing with a brand new client ID (i.e. in an incognito window) the first responses will make database queries but with each page load these will be replaced by cache hits. 
4. Verify the lightweight API works on an Atomic instance.
5. Verify the response speeds are faster than the REST API endpoints.

### Next steps:

1. Rewrite the `POST`endpoint in the same style.
2. Rework the`amp-access` pattern so the endpoint is hit only once even when there are multiple campaigns. This is the approach @adekbadek took in https://github.com/Automattic/newspack-audience
3. Auto-generate the config file when the plugin is installed.

### Other information:
